internal/servegit/service.go

```diff
@@ -0,0 +1,109 @@
+package servegit
+
+import (
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var uploadPackArgs = []string{
+	// Partial clones/fetches
+	"-c", "uploadpack.allowFilter=true",
+
+	// Can fetch any object. Used in case of race between a resolve ref and a
+	// fetch of a commit. Safe to do, since this is only used internally.
+	"-c", "uploadpack.allowAnySHA1InWant=true",
+
+	"upload-pack",
+
+	"--stateless-rpc",
+}
+
+// gitServiceHandler is a smart Git HTTP transfer protocol as documented at
+// https://www.git-scm.com/docs/http-protocol.
+//
+// This allows users to clone any git repo. We only support the smart
+// protocol. We aim to support modern git features such as protocol v2 to
+// minimize traffic.
+type gitServiceHandler struct {
+	// Dir is a funcion which takes a repository name and returns an absolute
+	// path to the filepath for it.
+	Dir   func(string) string
+	Debug *log.Logger
+}
+
+func (s *gitServiceHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Only support clones and fetches (git upload-pack). /info/refs sets the
+	// service field.
+	if svcQ := r.URL.Query().Get("service"); svcQ != "" && svcQ != "git-upload-pack" {
+		http.Error(w, "only support service git-upload-pack", http.StatusBadRequest)
+		return
+	}
+
+	var repo, svc string
+	for _, suffix := range []string{"/info/refs", "/git-upload-pack"} {
+		if strings.HasSuffix(r.URL.Path, suffix) {
+			svc = suffix
+			repo = strings.TrimSuffix(r.URL.Path, suffix)
+			repo = strings.TrimPrefix(repo, "/")
+			break
+		}
+	}
```

Comment:
Do we need a sanity check that `repo != "" && svc != ""`?

Reply:
empty repo is allowed (the root is a git repo), so don't wanna guard against it. We check for svc == "" later.